### ISSUE TITLE
Event when distributed Cache is updated

### ISSUE THREAD

**Issue by joshuacherkes** [MichaCo/CacheManager#93]
First Off: Thank you for this great product! it's worked wonders so far for me. 

I've been trying to do the following with cache manager, but havent figured out a way to do so: 

Lets say we have a client and server, each with an in memory cache backplaning to redis. 

If the server updates a key, the event OnUpdate will fire on the sever. 

Is it possible to also fire an event on the Client? the normal OnUpdate event does not fire in this scenario. I'm thinking of something along the lines of "OnUpdated". 

Thank You!

**Comment by TMiNus** [comment]
Just to clarify, you have a client and a server writing to a shared redis instance but with their own separated local caches?

If you update a key in the server machine you want that the client picking that change based on the OnUpdate event?

**Comment by joshuacherkes** [comment]
TMInus, 
That is correct, here is the cache configuration that both the client and the server use: 
```
            var cache = CacheFactory.Build<T>("cacheName", settings =>
            {
                settings
                    .WithSystemRuntimeCacheHandle("inProcessCache")
                    .And
                    .WithRedisConfiguration("redis", config =>
                    {
                        config.WithAllowAdmin()
                            .WithDatabase(0)
                            .WithEndpoint("localhost", 6379);
                    }).WithUpdateMode(CacheUpdateMode.Full)
                    .WithMaxRetries(1000)
                    .WithRetryTimeout(100)
                    .WithRedisBackplane("redis")
                    .WithRedisCacheHandle("redis", true);
            });
```

I would like it if the server changes a key, an event on the client is fired (Ideally within a given region).

**Comment by TMiNus** [comment]
With that configuration the client and the server will have their own redis instance, each one has a redis server listening on 6379. In the WithEndPoint you defined your local machine instead of an external one shared by both of them

**Comment by joshuacherkes** [comment]
Sorry I wasn't clear, I'm currently debugging and everything is running on localhost (client, server and redis).

**Comment by TMiNus** [comment]
Is this behavior only with the OnUpdate operation or with all the other events too?

**Comment by joshuacherkes** [comment]
I'm only interested in onupdate for the time being, but the other operations would also be nice. It seems that all of the events are localy firing. It does seem like in the RedisBackplane cache there are some events that might be useful. I'll take a look at those, but would appreciate anything else you can come up with.

**Comment by TMiNus** [comment]
Yes all the events start from *RedisBackPlane*, and *BaseCacheManager* subscribe to them to trigger his own events but as you can see there is no update event on the backplane just a changed event.

**Comment by TMiNus** [comment]
Another stuff I remember when I was playing with this part. A change in one key will actually trigger an eviction on the other versions of that specific key, which make sense because it is impossible to update every copy of that key, it is better to evict it in all the handles and let the client write the update key in all the handles

**Comment by TMiNus** [comment]
Well I went in :p **UpdateInternal** trigger the **OnUpdate** event but I think that is just local because the actual event that gets triggered in the BackPlane is **Changed**. Can you subscribe to the **Changed** event in the client, I am 98% it will pick the update operation

**Comment by joshuacherkes** [comment]
Just tested subscribing to the Changed event. This works nicely except that it fires any time ANY key is changed in redis! I think what this means for me practically is that I'll run seperate redis instances for each cache I want to operate. Is that a decent game plan or is there a better alternative? (the only other thing I can think of the just filtering the Changed event at the callback, but that can get cumbersome the more I use redis).

**Comment by TMiNus** [comment]
You can use the data of the event to filter which key was changed, look at **CacheItemEventArgs**

**Comment by TMiNus** [comment]
I will assume the code will go more or less like this

```
public void Example(object sender, CacheItemEventArgs args)
{
    var changedKey = args.Key;
    if (changedKey.Equals("YOUR FILTER KEY"))
        // Do something
}
```

Althought it can get cumbersome when you want to track multiple keys, you can always use a collection to put all the keys you wanna track and apply that to the method above with some LINQ magic

**Comment by MichaCo** [comment]
Currently for the backplane there is no difference if the item was updated or just changed via "Put" again.
I agree that it might be useful to have different events for that. But that would mean some changes in the backplane, another message etc...

Also, the events of the ICacheManager are not getting fired consistently by the backplane. Only `OnRemove` actually and `OnClear` gets called twice.

So overall, looks like I have to revisit the whole event stuff and maybe also add some more information ;)

**Comment by joshuacherkes** [comment]
Micah, 

Thanks for the response. I look forward to any improvements you make in the future!

**Comment by MichaCo** [comment]
This is fixed in 0.9.2, the events fired by BaseCacheManager are now also triggered through the backplane.
Event args now have an additional property `Origin` indicating if the event was triggered locally or remotely (through the backplane).

This actually works pretty nicely ;)

All events on the backplane itself are now internal, so if you've used those for some reason, you cannot do that anymore and there is no reason to. It was strange to get to anyways.

**Comment by MichaCo** [comment]
@joshuacherkes 
Version 0.9.2 is available now and and I changed a lot with the events (see release notes). Let me know if there is something not working or missing ;)
